### DISCUSSION TITLE
Not able to connect to Esera Ebus connector via Ethernet

### DISCUSSION THREAD

**Discussion by tobiaskb83** [john30/ebusd#837] [Q&A] [answered]
Hello community,

I am a very fresh user of the ebusd daemon. I want to read my heatings (Vaillant) ebus as the most of us here are doing it. Unfortunately I am struggling right in the beginning. Some facts first:

- rpi 4b with RaspbianGNU/Liunx 10 (buster) (Kernel: Linus 5.4.79-v7l+ / Architecture: arm) (ip 192.168.1.47:8888)
- esera ebus ethernet connector
- ip for connector is set and ebus connector visible in the LAN, ping from rpi is positive as well
- operation mode for connector is udp, IP is ....1.20:5000
![grafik](https://user-images.githubusercontent.com/73853572/217950599-e7da6ad7-3f6b-41fc-9551-e1775cf15082.png)

When I start the ebus this appears:
![grafik](https://user-images.githubusercontent.com/73853572/217950171-fd1655ee-39f8-4b70-babc-faa62cd9a81e.png)

I created rules on my unify dream machine for the ports and ips related to the data exchange.... nothing changes...

Ping from rpi as mentioned positive:
![grafik](https://user-images.githubusercontent.com/73853572/217950553-ba526598-e3ee-434f-934e-5d5d9620d38a.png)

I would be really happy to get a hint. Thanks in advance.

**Comment by Commifreak** [comment]
Maybe interesting: https://www.teunvandeberg.nl/posts/2021-06-07-connecting-a-vaillant-vrt350-thermostat-to-openhab/ (this was linked from here: https://github.com/john30/ebusd/wiki/6.-Hardware )

**Comment by tobiaskb83** [comment]
Thanks for the fast reply. I already came across that pages before. :-( Putty showed the same issues when I tried accessing the Esera adapter...

**Reply by Commifreak** [reply]
Maybe dumb, but: is it working in tcp mode?

**Reply by tobiaskb83** [reply]
Not dumb I think. :-) it is configured for udp. I xchecked that several times. Can be seen in the 1st Screenshot above. Cheers,  Tobias

**Comment by tobiaskb83** [comment]
I analyzed my network with the help of nmap and netstat and now I am confused. :-)

I can connect easily via netcat to UDP port 5000 of the Esera Ethernet Coupler, I can as well connect to UDP Port 8888 on my rpi. I can see those connections as active ones with netstat -u. 

In the wiki I read that you can as well connect with netcat to 8888 to listen to the ebus. But why do I have to connect still to the Ebus Coupler when it sends via udp from Port 5000 with its IP to the RPI UDP Port 8888 and its IP? I guess I don´t get what the black box does exactly. :-) It is still the case that this error is thrown, despite the fact I can reach the coupler via netcat:
![grafik](https://user-images.githubusercontent.com/73853572/218223908-dab59e2f-2752-451b-8354-bdad4ef7db8d.png)

**Comment by nuki47** [comment]
Same situation, I tried
`ebusd -f -c /tmp --logareas bus --loglevel info --lograwdata=bytes -d 192.168.200.238:5000`
an I became unable to open...
but I can ping the ebus coupler and ebusctl i bring this feedback
![grafik](https://user-images.githubusercontent.com/75117725/218269382-a333aa71-bc27-4a8a-88d5-66c99d7f3613.png)

**Comment by kjumybit** [comment]
Hi,
I got the same problem using UDP.
Please verify, that you are using the most recent ebusd version "23.1" or the current master.
(see #733)

**Reply by tobiaskb83** [reply]
Hey. Thanks for the hint. Will try it tonight. What do you mean with master? Ebusd is current built w/ 23.1.

**Reply by kjumybit** [reply]
Hi, 
you can clone the current version (last commit) of the source from the 'master' branche' and build the ebusd binary by your own. John30 builds binaries with an official version number periodically. the issue above is fixed in version 23.1.

**Reply by tobiaskb83** [reply]
I think I got it. But my ebusd is already 23.1. I'll try your solution nontheless. I think I understood what you said. Thx.

**Comment by nuki47** [comment]
whats happen here?
![grafik](https://user-images.githubusercontent.com/75117725/218316994-b5616ab2-44ef-458f-a4eb-f808660d9bad.png)

![grafik](https://user-images.githubusercontent.com/75117725/218316915-c8e40240-6b19-45d5-9db8-242c4ba6461a.png)

**Reply by kjumybit** [reply]
Hi,
please use an alternative approach for scanning the UDP connection.

`nmap -sU -v 192.168.200.238`

After a some time the result should show an open port e.g. on my Esera adapter:

'5000/udp open|filtered upnp'

full log:
```
Starting Nmap 7.93 ( https://nmap.org ) at 2023-02-12 17:57 CET
Initiating ARP Ping Scan at 17:57
Scanning 192.168.x.x [1 port]
Completed ARP Ping Scan at 17:57, 0.06s elapsed (1 total hosts)
Initiating Parallel DNS resolution of 1 host. at 17:57
Completed Parallel DNS resolution of 1 host. at 17:57, 0.00s elapsed
Initiating UDP Scan at 17:57
Scanning 192.168.x.x [1000 ports]
I...
UDP Scan Timing: About 86.70% done; ETC: 18:11 (0:01:52 remaining)
UDP Scan Timing: About 91.80% done; ETC: 18:11 (0:01:09 remaining)
Completed UDP Scan at 18:12, 883.08s elapsed (1000 total ports)
Nmap scan report for 192.168.x.x
Host is up (0.0031s latency).
Not shown: 999 closed udp ports (port-unreach)
PORT     STATE         SERVICE
5000/udp open|filtered upnp
MAC Address:XX:XX:XX:XX:XX:XX (Wiznet)

Read data files from: /snap/nmap/2944/usr/bin/../share/nmap
Nmap done: 1 IP address (1 host up) scanned in 883.27 seconds
           Raw packets sent: 1191 (54.495KB) | Rcvd: 1183 (66.220KB)
```

**Reply by nuki47** [reply]
what's wrong here?
```
Starting Nmap 7.80 ( https://nmap.org ) at 2023-02-12 19:34 CET
Initiating ARP Ping Scan at 19:34
Scanning 192.168.200.238 [1 port]
Completed ARP Ping Scan at 19:34, 0.00s elapsed (1 total hosts)
Initiating Parallel DNS resolution of 1 host. at 19:34
Completed Parallel DNS resolution of 1 host. at 19:34, 0.03s elapsed
Initiating UDP Scan at 19:34
Scanning 192.168.200.238 [1000 ports]
Increasing send delay for 192.168.200.238 from 0 to 50 due to max_successful_tryno increase to 4
Increasing send delay for 192.168.200.238 from 50 to 100 due to max_successful_tryno increase to 5
UDP Scan Timing: About 44.40% done; ETC: 19:35 (0:00:39 remaining)
Completed UDP Scan at 19:35, 88.19s elapsed (1000 total ports)
Nmap scan report for 192.168.200.238
Host is up (0.00055s latency).
All 1000 scanned ports on 192.168.200.238 are closed
MAC Address: 00:08:DC:53:8C:16 (Wiznet)

Read data files from: /usr/bin/../share/nmap
Nmap done: 1 IP address (1 host up) scanned in 88.40 seconds
           Raw packets sent: 1112 (32.071KB) | Rcvd: 1112 (62.244KB)
```

**Reply by kjumybit** [reply]
Hi @nuki47,
there are to few information for a root cause analysis. It would start to review the Esera adapter configuration within the config tool. Can you post a screen shot?

**Reply by nuki47** [reply]
do you mean that?
![grafik](https://user-images.githubusercontent.com/75117725/218545423-83e1e886-7130-41ed-b06f-ed2bfea7ca64.png)

**Reply by kjumybit** [reply]
Yes, that's it.

You configured your adapter to listen on a TCP port, you enabled `TCP server`, but you start your ebusd with an udp protocol option `udp:192.168.200.238:5000`. 

That is not consistent.  

I use UDP because it should have a lower latency. To use UDP please check the radio button `UDP` in the config tool an push the configuration to the device using the button `Setting`.

After that try again, e.g. use nmap to scan for the open UDP port first or start the ebusd immediately.

**Reply by nuki47** [reply]
Step 1)
![grafik](https://user-images.githubusercontent.com/75117725/218564925-723fec5f-93ff-46c1-895b-2b39b6bc2f94.png)

Step 2)
```
pi@raspberry:~ $ nmap 192.168.200.238
Starting Nmap 7.80 ( https://nmap.org ) at 2023-02-13 21:14 CET
Nmap scan report for 192.168.200.238
Host is up (0.0012s latency).
All 1000 scanned ports on 192.168.200.238 are closed
```
:-(

**Reply by kjumybit** [reply]
Your `Remote IP/host name` has the same IP address as the device. That's wrong. The remote host is the host, that runs the ebusd. Please use the the IP address of your Raspberry Pi. The device IP address is a new IP address, which you assign to your Esera ebus adapter. That address must not be used in your network.

**Reply by nuki47** [reply]
so, now I get:
```
pi@raspberry:~ $ nmap 192.168.200.238
Starting Nmap 7.80 ( https://nmap.org ) at 2023-02-14 06:10 CET
Nmap scan report for 192.168.200.238
Host is up (0.0015s latency).
Not shown: 999 closed ports
PORT      STATE SERVICE
```
![grafik](https://user-images.githubusercontent.com/75117725/218645050-bd21e16f-417d-4fe3-a883-cc6492981fa6.png)

**Reply by kjumybit** [reply]
Hi, please check with `sudo nmap -sU -v 192.168.200.238`.

**Reply by nuki47** [reply]
```
pi@raspberry:~ $ sudo nmap -sU -v 192.168.200.238
Starting Nmap 7.80 ( https://nmap.org ) at 2023-02-14 20:58 CET
Initiating ARP Ping Scan at 20:58
Scanning 192.168.200.238 [1 port]
Completed ARP Ping Scan at 20:58, 0.00s elapsed (1 total hosts)
Initiating Parallel DNS resolution of 1 host. at 20:58
Completed Parallel DNS resolution of 1 host. at 20:58, 0.07s elapsed
Initiating UDP Scan at 20:58
Scanning 192.168.200.238 [1000 ports]
Increasing send delay for 192.168.200.238 from 0 to 50 due to max_successful_tryno increase to 4
Increasing send delay for 192.168.200.238 from 50 to 100 due to max_successful_tryno increase to 5
UDP Scan Timing: About 39.77% done; ETC: 20:59 (0:00:47 remaining)
Completed UDP Scan at 21:00, 95.38s elapsed (1000 total ports)
Nmap scan report for 192.168.200.238
Host is up (0.0015s latency).
Not shown: 999 closed ports
PORT     STATE         SERVICE
5000/udp open|filtered upnp
MAC Address: 00:08:DC:53:8C:16 (Wiznet)
```

**Reply by nuki47** [reply]
```
pi@raspberry:~ $ sudo ebusd -f -c /tmp --logareas bus --loglevel info --lograwdata=bytes -d udp:192.168.200.238:5000
2023-02-14 21:02:17.333 [bus notice] bus started with own address 31/36
2023-02-14 21:02:17.350 [bus notice] <aa
2023-02-14 21:02:17.350 [bus notice] signal acquired
2023-02-14 21:02:17.398 [bus notice] <aa
2023-02-14 21:02:17.444 [bus notice] <aa
2023-02-14 21:02:17.491 [bus notice] <aa
2023-02-14 21:02:17.538 [bus notice] <aa
```

**Reply by nuki47** [reply]
@kjumybit What do you think? do you have some more hints?

**Reply by kjumybit** [reply]
Hi,
your ebusd is connected to the adapter and receives messages. Now it's up to you to complete the configuration. 

Please read the Run section of the [wiki](https://github.com/john30/ebusd/wiki/2.-Run)

- provide the message definition files you want to use, get it from [here](https://github.com/john30/ebusd-configuration)
- define the runtime options
- check the message processing using the `ebusctl` command line tool
- run the ebusd as system service
- setup an MQTT broker to integrate the ebusd into your  preferred home automation controller , e.g. Homeassistant

Here is an example for ebusd  with few basic options I use:  
- upd protocol
- with a local copy of the configuration files
- without MQTT

```
BUSD_OPTS="--configpath=/etc/ebusd --scanconfig --device udp:192.168.2.19:5000 --initsend --latency=2000 --nodevicecheck --updatecheck=off
```
you ca use the `ebusctl` tool to check for correct message detection of your devices.
On your ebusd host run:

```
ebusctl 
```
enter `info` to get an enhanced status report.  Here you can see my Vaillant heating devices and loaded confguration files.
The value of the `signal` attribute must be `acquired`, otherwise the signal is not clearly detected by the Esera adapter and the adapter must be trimmed (see the manual for that).
```
localhost: info
version: ebusd 22.4.v22.4-57-g5d3c6f09
device: udp:192.168.2.19:5000
signal: acquired
symbol rate: 68
max symbol rate: 114
min arbitration micros: 1
max arbitration micros: 163
min symbol latency: 7
max symbol latency: 40
reconnects: 0
masters: 3
messages: 760
conditional: 9
poll: 4
update: 11
address 03: master #11
address 08: slave #11, scanned "MF=Vaillant;ID=BAI00;SW=0414;HW=7401", loaded "vaillant/bai.0010006101.inc" ([PROD='']), "vaillant/08.bai.csv"
address 10: master #2
address 15: slave #2, scanned "MF=Vaillant;ID=UI   ;SW=0324;HW=6201", loaded "vaillant/15.ui.csv"
address 23: slave, scanned "MF=Vaillant;ID=SOLSY;SW=0306;HW=6301", loaded "vaillant/23.solsy.cc.csv"
address 25: slave, scanned "MF=Vaillant;ID=SOLSY;SW=0306;HW=6301", loaded "vaillant/25.solsy.hwc.csv"
address 26: slave, scanned "MF=Vaillant;ID=SOLSY;SW=0306;HW=6301", loaded "vaillant/26.solsy.hc.csv"
address 31: master #8, ebusd
address 36: slave #8, ebusd
address 50: slave, scanned "MF=Vaillant;ID=SOLSY;SW=0306;HW=6301", loaded "vaillant/50.solsy.mc.csv"
address ec: slave, scanned "MF=Vaillant;ID=SOLSY;SW=0306;HW=6301", loaded "vaillant/ec.solsy.sc.csv"
```

**Reply by nuki47** [reply]
ok, but `ebusctl` don't bring much data?
```
localhost: info
version: ebusd 23.1.23.1
device: udp:192.168.200.238:5000
signal: acquired
symbol rate: 44
max symbol rate: 89
min arbitration micros: 4
max arbitration micros: 22
min symbol latency: 10
max symbol latency: 20
reconnects: 0
masters: 4
messages: 3
conditional: 0
poll: 0
update: 0
address 01: master #6
address 03: master #11
address 06: slave #6, scanned "MF=TEM;ID=MB61;SW=0044;HW=0000"
address 08: slave #11, scanned "MF=TEM;ID=WE_1 ;SW=3632;HW=3030"
address 10: master #2
address 15: slave #2, scanned "MF=TEM;ID=19469;SW=0359;HW=0110"
address 31: master #8, ebusd
address 36: slave #8, ebusd
```

**Reply by kjumybit** [reply]
Hi @nuki47 ,
my recommendation to you is to read the wiki, all sections. You need a basic understanding, how ebusd works, what a ebus message is and how the tools like ebusctl can help you, discover the messages. This thread deals with the connection to the esera adapter. That seems to be working in your case.
Please open a new discussion and ask a concrete question. Provide the steps you have already done and detailed  information of your results and local configuration.

**Comment by kjumybit** [comment, marked as answer]
Hi @tobiaskb83,

the screenshot of your Esera config tool shows a remote host port `8888`, which differs from the default `5000`. Actually, I don't know, why it is necessary to specify a client port. In my configuration I don't change that port and can't even do that. After saving the new config  via button "Setting" and re-open the config tool, the port value is always `5000`. 

So maybe restoring the same (the default) UDP port as specified for the adapter would help.

On your ebusd host, the local UDP port must not be already in use by any other program, e.g. an already running (old) instance of the ebusd.

To verify that run

`netstat -u`

For me it shows my running ebusd.

```
~$ netstat -u
Active Internet connections (w/o servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State      
udp        0      0 up2one:5000             192.168.2.19:5000       ESTABLISHED
```

**Reply by tobiaskb83** [reply]
Hi @kjumybit, I looked into your proposal and that is what happend:
![grafik](https://user-images.githubusercontent.com/73853572/218588333-46b61c12-2c22-4133-857f-639b259000f4.png)

If I start ebusd after rpi restart the bus starts with "own address". After killing the running ebusd at least the udp ports of the esera and the rpi are connected. Guess ebusd did that. Before ebusd started they weren´t. If I then I start the bus again it says again ERR: generic I/O error.

I changed the remote ports as you suggested to 5000. That seems to work. But there seems still an issue with the connection process. :-/

Config for esera looks like this now:
![grafik](https://user-images.githubusercontent.com/73853572/218589124-f02a3afe-c4a7-4bab-a8e1-6bcc78ea6727.png)

At least a small devolopment. Thanks @kjumybit.
Tobias

**Reply by tobiaskb83** [reply]
@kjumybit FYI: my bus is running now. Your hint was key to get it running. Thanks a lot. If I want to kill ebusd in foreground and launch it again. I simply have to kill the established udp port connection via terminal commands. Then I can run again the bus.

Thanks a lot. Best regards. Tobias

**Comment by nuki47** [comment]
I would come back to this thread because I have some problems to connect again.
`sudo nmap -sU -v 192.168.200.238` bring this result:
```
PORT     STATE         SERVICE
5000/udp open|filtered upnp
```
but `ebusctl i` brings this result:
```
version: ebusd 23.1.23.1
update check: OK
device: udp:192.168.200.238:5000, invalid
signal: no signal
reconnects: 0
masters: 1
messages: 0
conditional: 0
poll: 0
update: 0
address 31: master #8, ebusd
address 36: slave #8, ebusd
```
I also switched the heat pump and ebus controller off/on before.

![grafik](https://user-images.githubusercontent.com/75117725/221406221-e39bde8d-1428-4d8d-a1ae-bac4028574a6.png)

`sudo ebusd -f -c /tmp --logareas bus --loglevel info --lograwdata=bytes -d udp:192.168.200.238:5000`
bring this result:

```
2023-02-26 12:04:53.672 [bus error] unable to open udp:192.168.200.238:5000: ERR: generic I/O error
2023-02-26 12:04:53.673 [bus notice] bus started with own address 31/36
2023-02-26 12:04:53.673 [bus notice] device invalid
2023-02-26 12:04:58.673 [bus error] unable to open udp:192.168.200.238:5000: ERR: generic I/O error
2023-02-26 12:04:58.673 [bus notice] device invalid
2023-02-26 12:05:03.674 [bus error] unable to open udp:192.168.200.238:5000: ERR: generic I/O error
2023-02-26 12:05:03.674 [bus notice] device invalid
```
